db/db_impl/db_impl_write.cc

```diff
@@ -1869,27 +1868,26 @@ Status DBImpl::DelayWrite(uint64_t num_bytes, WriteThread& write_thread,
   return s;
 }
 
-// REQUIRES: mutex_ is held
 // REQUIRES: this thread is currently at the front of the writer queue
-void DBImpl::WriteBufferManagerStallWrites() {
-  mutex_.AssertHeld();
-  // First block future writer threads who want to add themselves to the queue
-  // of WriteThread.
-  write_thread_.BeginWriteStall();
-  mutex_.Unlock();
-
-  // Change the state to State::Blocked.
-  static_cast<WBMStallInterface*>(wbm_stall_.get())
-      ->SetState(WBMStallInterface::State::BLOCKED);
-  // Then WriteBufferManager will add DB instance to its queue
-  // and block this thread by calling WBMStallInterface::Block().
-  write_buffer_manager_->BeginWriteStall(wbm_stall_.get());
-  wbm_stall_->Block();
-
-  mutex_.Lock();
-  // Stall has ended. Signal writer threads so that they can add
-  // themselves to the WriteThread queue for writes.
-  write_thread_.EndWriteStall();
+void DBImpl::MaybeWriteBufferManagerStallWrites() {
+  bool begin_stall =
+      write_buffer_manager_->MaybeBeginWriteStall(wbm_stall_.get());
+  if (begin_stall) {
+    mutex_.Lock();
+    // Block future writer threads who want to add themselves to the queue
+    // of WriteThread.
+    write_thread_.BeginWriteStall();
```

Comment:
Question 1: Why do want to block future writer threads from being added to the queue of WriteThread, even though adding to the queue does not mean "getting to execute"? 

This writer thread is currently at the front of the writer queue, which IIUC, is the only writer thread that can do the real write work.  With this front thread being blocked due to write stall in the next logic, all threads behind it in the queue are technically blocked from doing the real write work. From this perspective, it seems fine to me to let future write threads to add themselves to the queue but they will just have to wait in the queue though.

This is somehow related to my optimization 1's concern below.

Reply:
> Why do want to block future writer threads from being added to the queue of WriteThread, even though adding to the queue does not mean "getting to execute"?


Actually let me run it without the "block" and see what weird stuff pop up :)

Reply:
The question is not urgent as we don't have to get it answered to meet the current need (SetAllowStall())